plugin/cache/handler.go

```diff
@@ -81,6 +70,13 @@ func (c *Cache) doPrefetch(ctx context.Context, state request.Request, server st
 	}
 }
 
+func (c *Cache) doRefresh(ctx context.Context, state request.Request, cw *ResponseWriter) (int, error) {
+	if !state.Do() {
+		setDo(state.Req)
```

Comment:
So, now we have this operating on the state.Req which is a pointer to a dns.Msg. Is this thread safe? It's still done in a go function? So I have some trouble seeing how this change would now fix the race?

Reply:
there is no race because no one else is accessing `state.Req`, we could move all prefetches after response is written to make sure no one really uses state.Req, but this should already be true

Reply:
or we could modify the request on the main thread and then fire a goroutine